### PR TITLE
New code docs urls

### PR DESCRIPTION
**Pull request by bethanyaconnor**
This PR adds usable URL for CSA docs and new, more Rails-y, URLs for existing code docs. There's more details in https://docs.google.com/document/d/1aZLFRXceY7np9Yzi2JtkXuAWDs0EANAel627YXjqL8I/edit. 

For programming classes, these URLs are new and the ones we will use going forward. For existing docs, once we launch, we will redirect from the old URLs to the new ones. I tested this by going directly to the page, viewing a doc in the "Help and Tips" tab, and viewing a doc from the IDE by clicking on the link from the blocks in applab and gamelab. All three worked without issue. Note that this won't be live until we flip the `use_code_studio_docs` DCDO flag.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Review by molly-moen — APPROVED** [review]
LGTM!

**Comment by tess323** [comment]
This works for me.  Thank you for all the deep thought here.

File: dashboard/app/controllers/programming_classes_controller.rb [hunk 1]
```diff
@@ -1,6 +1,7 @@
 class ProgrammingClassesController < ApplicationController
-  before_action :require_levelbuilder_mode_or_test_env
+  before_action :require_levelbuilder_mode_or_test_env, only: [:new, :create, :edit, :update]
   load_and_authorize_resource
+  before_action :set_class_by_keys, only: [:show_by_keys]
```
Comment (jamescodeorg) [inline dashboard/app/controllers/programming_classes_controller.rb]: Mostly curious, why did you specify this as a `before_action` instead of just calling it from `show_by_keys`?

File: dashboard/app/controllers/programming_classes_controller.rb [hunk 2]
```diff
@@ -58,6 +59,15 @@ def show
     @programming_environment_categories = @programming_class.programming_environment.categories_for_navigation
   end
 
+  def show_by_keys
+    return render :not_found unless @programming_class
+    if params[:programming_environment_name] && params[:programming_class_key]
```
Comment (jamescodeorg) [inline dashboard/app/controllers/programming_classes_controller.rb]: Does this check guarantee that `@programming_class` is non-nil?  Would it be more appropriate to check `if @programming_class` here instead?

File: dashboard/app/controllers/programming_expressions_controller.rb [hunk 3]
```diff
@@ -137,8 +137,7 @@ def clone
   def docs_show
     if DCDO.get('use-studio-code-docs', false)
       return render :not_found unless @programming_expression
-      @programming_environment_categories = @programming_expression.programming_environment.categories_for_navigation
-      return render :show
+      return redirect_to(programming_environment_programming_expression_path(@programming_expression.programming_environment.name, @programming_expression.key))
```
Comment (jamescodeorg) [inline dashboard/app/controllers/programming_expressions_controller.rb:138]: I forget, do we need a `return` before `redirect_to`?  (I see we have a `return` here but not up in `programming_environments_controller.rb`.)

File: dashboard/app/controllers/programming_classes_controller.rb [hunk 4]
```diff
@@ -58,6 +59,15 @@ def show
     @programming_environment_categories = @programming_class.programming_environment.categories_for_navigation
   end
 
+  def show_by_keys
```
Comment (jamescodeorg) [inline dashboard/app/controllers/programming_classes_controller.rb:85]: Could we add a comment with the URL that gets routed here (like we do in some other parts of the code)?  It'd be nice if we could find this method by searching for the beginning of the URL.

Same for ProgrammingExpression.  Also ok to do this in a separate PR if it feels unrelated.

**Review by jamescodeorg — APPROVED** [review]
LGTM!

Could we update the url doc to say that we picked option 1 and that we're using "expression" instead of "code" in the url?

File: dashboard/app/controllers/programming_classes_controller.rb (hunk 2 above)
Comment (bethanyaconnor) [reply dashboard/app/controllers/programming_classes_controller.rb]: Yeah, I think this is redundant with the lines above plus `set_class_by_keys`. I'll remove!

File: dashboard/app/controllers/programming_expressions_controller.rb (hunk 3 above)
Comment (bethanyaconnor) [reply dashboard/app/controllers/programming_expressions_controller.rb:138]: I apparently used slightly different patterns 🙃 programming_environments_controller uses an if/else so it doesn't need a return, whereas this controller action is not. I'll pick one and standardize the two

File: dashboard/app/controllers/programming_classes_controller.rb (hunk 4 above)
Comment (bethanyaconnor) [reply dashboard/app/controllers/programming_classes_controller.rb:85]: I'll add for these actions in this PR as they're new but it would be good for me to add it for all in a separate PR

File: dashboard/app/controllers/programming_classes_controller.rb (hunk 1 above)
Comment (bethanyaconnor) [reply dashboard/app/controllers/programming_classes_controller.rb]: I think ideally, it should play with `authorize_resource` (might need to rearrange though). But I couldn't get that to work so I'm moving all of it into the controller action for now.

This did make me realize I forgot some tests though...
